src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -4129,24 +4130,26 @@ public void clearSnapshot(String tag, String... keyspaceNames) throws IOExceptio
 
     public Map<String, TabularData> getSnapshotDetails(Map<String, String> options)
     {
+        boolean skipExpiring = options != null && Boolean.parseBoolean(options.getOrDefault("no_ttl", "false"));
+
+        SnapshotLoader loader = new SnapshotLoader();
         Map<String, TabularData> snapshotMap = new HashMap<>();
-        for (Keyspace keyspace : Keyspace.all())
+
+        for (TableSnapshot snapshot : loader.loadSnapshots())
```

Comment:
I see you have 'walked' variable in loadSnapshots method, I guess it is for not needing to go over whole data structure multiple times. However, is it really used in this method? This is called from JMX when I am not mistaken, plus you create new instance of SnapshotLoader every time we use nodetool, so it will parse all data dirs anyway. Maybe we should cache this? I think it would be better to do this via SnapshotManager and I believe you have this in mind, just the proximity of the release is blocking you from doing it "properly".

Reply:
> I see you have 'walked' variable in loadSnapshots method, I guess it is for not needing to go over whole data structure multiple times. However, is it really used in this method?

Removed `walked` variable from `SnapshotLoader` since it's no longer neeeded. This is a leftover from a previous implementation.


> Maybe we should cache this? I think it would be better to do this via SnapshotManager and I believe you have this in mind, just the proximity of the release is blocking you from doing it "properly".

Yes, I have a longer version of this patch where loaded snapshots are cached in the `SnapshotManager` ([here](https://github.com/apache/cassandra/compare/trunk...pauloricardomg:CASSANDRA-16843-final?expand=1#diff-17b12c3f18d24601e8d6dcdf288f6583624ebf1b5cd1b135cfe1ed5abd6d349dR71)). However in order to cache this properly we need to move snapshot cleanup logic to `SnapshotManager`, which makes this patch bigger than it should be and riskier for 4.1. I plan to create follow-up patches based on the longer patch above centralizing snapshot logic on `SnapshotManager` for 4.2.